### ISSUE TITLE
Slowness on Devices tab UI

### ISSUE THREAD

**Issue by charlyforot** [librenms/docker#343]
Hello,

We recently migrated LibreNMS on docker and we found a huge slowness on the "devices" tab.

When clicking on the page, we saw with htop a lot of top rrdcached processes (graph loading).
The page then takes 20 seconds to load and we suspect that this is due to the loading of the graphs.
It seems that devices with many ports take a long time to load.

As a result, we have to wait until the page has finished loading before we can click on the various filters on the page.

Without docker, LibreNMS UI has not this slowness.
We have tried to increase PHP settings and opcache values but there were no improvements.

I specify that the previous installation was based on Apache whereas on the new container there is Nginx.

Somes things I tried :

- Increase PHP settings (memory_limit, max_children, opcache)
- Increase somes Nginx conf values
- Disable 'enable_lazy_load' of LibreNMS config

Some similar topics :

- https://github.com/librenms/librenms/issues/14021#issuecomment-1149909759
- https://community.librenms.org/t/librenms-is-loading-very-slow-despite-optimizations/303/19

None of these changes made an improvement.

Do you have an idea to solve this problem ?

Thank you !

#### Steps to reproduce this issue

1. Have a lot of ports on some devices
2. Click on "Devices" tab on UI

#### Expected behaviour

The page must load in about a second.

#### Actual behaviour

Notice that the page takes a long time to load (20 seconds).
We can't apply a filter during this time.

### Configuration

* Docker version (type `docker --version`) : Docker version 20.10.6, build 370c289
* Platform (Debian 9, Ubuntu 18.04, ...) : Debian 10.9
* env

```
MEMORY_LIMIT=1024M
MAX_INPUT_VARS=1000
UPLOAD_MAX_SIZE=16M
OPCACHE_MEM_SIZE=256
REAL_IP_FROM=XXXXXXXXX
REAL_IP_HEADER=X-Forwarded-For
LOG_IP_VAR=remote_addr
CACHE_DRIVER=redis
SESSION_DRIVER=redis
REDIS_HOST=XXXXXXXXX
REDIS_PORT=XXXXXXXXX
REDIS_PASSWORD=XXXXXXXXX
LIBRENMS_SNMP_COMMUNITY=XXXXXXXXX
RRDCACHED_SERVER=XXXXXXXXX
MEMCACHED_HOST=XXXXXXXXX
MEMCACHED_PORT=XXXXXXXXX
FPM_PM_MAX_CHILDREN=50
FPM_PM_MAX_SPARE_SERVERS=11
FPM_PM_START_SERVERS=10
```

### Docker info

```
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Build with BuildKit (Docker Inc., v0.5.1-docker)
  scan: Docker Scan (Docker Inc., v0.7.0)

 Kernel Version: 4.19.0-16-amd64
 Operating System: Debian GNU/Linux 10 (buster)
 OSType: linux
 Architecture: x86_64
 CPUs: 24
 Total Memory: 125.4GiB
```

**Comment by murrant** [comment]
Yes, the os population is slow because the SQL table doesn't have an index for that query.

**Comment by charlyforot** [comment]
> Yes, the os and types selection population is slow because the SQL table doesn't have an index for that query.

Hello Murrant,

It is the same problem with all others filters. Moreover, our old LibreNMS install has not this slowness on this page.
Could it be due to the difference in http server (apache/nginx) ?

**Comment by murrant** [comment]
Disabling lazy load would have the opposite affect than you want.  Lazy load causes the graphs to be loaded after the page, instead of stopping the page load to wait for every graph.

**Comment by murrant** [comment]
@charlyforot If you want to get a little more info do the following:

**CAUTION! this is insecure, don't leave APP_DEBUG enabled!**

1. Install developer packages `./scripts/composer_wrapper.php install` (omitting --no-dev)
2. Enable Debug by setting APP_DEBUG=true in .env (or just in the environment)
3. Run `php artisan optimize` to update some cached config variables
4. Visit the page an open the debug bar at the bottom of the page
5. Make sure to click the "folder" icon to view different requests the page makes.

You might want to look at the webdev developers tab to identify any slow requests.

6. Disable APP_DEBUG (other stuff doesn't matter and it will be removed by daily)

**Comment by UrsaSolar** [comment]
Throwing my findings in here for posterity:

We've been having this issue too, almost identical to the letter other than several times longer load, and we went through all the same troubleshooting steps before finding this post. As a last ditch effort we decided to just kill the rrdcached container entirely. This caused an increase of write IOPS of about 400-600% on a reasonably large site (about 370 devices with about 20,000 ports resulted in going from ~400IOPS to ~2000 IOPS with some variance and spiking) but thankfully we have enough overhead room to just eat that cost. This may not be acceptable for other workloads or hardware however, I'm not really well versed in IOPS scaling.

We're only about a day into testing it, but it otherwise seems to work perfectly fine. The web GUI now responds extremely quickly and no longer locks the system up for minutes at a time as rrdcached struggles to respond. Like @charlyforot we didn't have this issue with a previous native install, though ours was still using nginx. This seems to be something specific to rrdcached running as a separate container being unable to serve data efficiently.

On Discord it was mentioned it is possible to load the RRD files directly from the core container, but then they'll be out of date by however long it was since the last flush. This could be acceptable as long as the flush time was shortened, and we may give that a shot later, but for the short term we're sticking with just not using rrdcached since this is finally working for us.
